design/ozone.md

```diff
@@ -0,0 +1,246 @@
+ * The design is based on the HDFS design. It can be scaled easily to thousand of nodes (but in the mean time the legendary small file problem of HDFS is also solved.)
+
+
+## Components of Ozone
+
+Ozone has the following components in a typical cluster:
+
+ * Two kind of metadata servers:
+    * **Storage Container Manager** (SCM): for block space management (create huge binary blocks which are replicated between the datanodes)
+    * **Ozone Manager** (OM): for key space management (create buckets, keys on top the blocks which are replicated by the Storage Container Manager)
+ * **Datanodes**: Storage daemons
+ * **S3 gatway** (stateless rest services, with AWS S3 compatible interface)
+ * **Recon** server (generic UI, anomaly detection, historical data)
+ * DB backend for recon
+
+The CSI interface requires two additional daemons:
+
+  * CSI Controller service
+  * CSI Node service
+  
+Ozone uses `StatefulSet` for the master servers (Storage Container Manager, Ozone Manager, S3 gateway, Recon Server) and `DaemonSet` for datanodes
+
+![Ozone components](https://i.imgur.com/e7Tmij4.png)
+
+In a real world setup Ozone use the following components:
+
+ * Prometheus to collect all of the metrics
+ * Grafana to display the default dashboards
+
+## Storage requirements
+
+Similar to other Software Defined Storage, Ozone is the gateway between the local storage and network storage. 
+
+It's suggested to use local storage. Ozone suggests to use SSD or NVMe, for the metadata store and spinning disk for the real data.
+
+## Implementation
+
+The initial implementation of the Operator should be based on the available examples. The main beahviour will be similar to the minio operator (alpha), it will install Ozone cluster based on a custom CRD. Some notable differences.
+
+ * Versioning support (see the next section)
+ * Kubernetes resources are generated from the original k8s template files provided by the Ozone project (it is an advanced version of the existing ceph/csi example)
+
+### Version support
+
+By default, Ozone tools are not required by the operator. The only reason is to use the `apache/ozone` container image as the base image is to have all the required k8s resources templates.
+
+The version of Ozone can be decoupled from the version of the Operator similar to the [CSI version decoupling](https://github.com/rook/rook/blob/master/design/decouple-ceph-version.md). The default version will be defined by the rook operator version but other version can be used based on CRD entry. 
+
+### Kubernetes resource generation
+
+Ozone provides examples K8s resources files and transformation definitions using [flekszible](https://github.com/elek/flekszible). This is a composition based k8s resource generator where the transformations can be reused.
+
+Let's say we have a simple kubernetes resource:
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: datanode
```

Comment:
Is the SCM the backing store for the data node? How does the data node connect to the SCM? It's not a PVC that would be attached to a statefulset?

Reply:
SCM is one of the metadata/master server. We have two metadata servers:

 1. SCM handles the mapping between huge binary blocks and the datanodes (the location of the huge blocks)
 2. OzoneManager stores the metadata for the object store semantic (buckets, keys) and under the hood it uses the replication capability of SCM. (Mapping between key -> huge binary block id)

Datanodes are connecting to the SCM and reporting the available blocks with a heartbeat mechanism (it's very similar to the existing HDFS approach).

The last question is very good. Until now we supported both DaemonSet and StatefulSet based datanode definitions. But recently we found it that it would be easier to use StatefulSet for the datanodes because the volumeClaimTemplate requirements.

From HDDS-1646 (apache/hadoop#945) we will support only StatefulSets for the datanode.

I updated this example in the design doc and an other paragraph (where we mentioned which k8s resources are used for Ozone objects.